src/buffer_research.py

```diff
@@ -0,0 +1,225 @@
+from src.oba_from_uniswap.common import get_block_data_file, get_prices_at_blocks
+import pickle
+
+from src.dune_api.dune_analytics import DuneAnalytics
+import pandas as pd
+from math import log
+import csv
+
+
+pd.set_option('display.max_rows', 50)
+pd.set_option("display.min_rows", 49)
+
+TOL = 1e-7
+
+total_matched_vol = 0
+total_unmatched_vol = 0
+total_rebalanced_vol = 0
+
+
+def count_number_of_saved_trades_due_to_cow(df):
+    number_of_trades_per_pair = df.groupby(
+        ["token_a_address", "token_b_address"]).size()
+    number_of_saved_trades_per_pair = dict()
+    for tf, count in number_of_trades_per_pair.items():
+        t, f = tf
+        if t < f:
+            t, f = f, t
+        if (t, f) in number_of_saved_trades_per_pair:
+            number_of_saved_trades_per_pair[t, f] += count
+        else:
+            number_of_saved_trades_per_pair[t, f] = count - 1
+    return sum(
+        number_of_saved_trades_per_pair.values())
+
+
+def apply_batch_trades_on_buffer_and_account_trade_statistic(sent_volume_per_pair, buffers, buffer_allow_listed_tokens):
+    nr_of_external_trades = 0
+    nr_of_internal_trades = 0
+    sum_rebalance_vol = 0
+    sum_matched_vol = 0
+    previous_buffer_usd_value = sum(buffers.values())
+
+    pair_visited = {}
+    for (f, t) in sent_volume_per_pair.keys():
+        # don't visit a pair twice
+        if (f, t) in pair_visited:
+            continue
+        else:
+            pair_visited[f, t] = True
+            pair_visited[t, f] = True
+
+        # calculate the volumes sent and received on pair
+        sent_vol_from_to = sent_volume_per_pair[f, t]
+        if not (t, f) in sent_volume_per_pair:
+            sent_vol_to_from = 0
+        else:
+            sent_vol_to_from = sent_volume_per_pair[t, f]
+        if sent_vol_from_to < sent_vol_to_from:
+            f, t = t, f
+            sent_vol_from_to, sent_vol_to_from = sent_vol_to_from, sent_vol_from_to
+        # calculate the matched amounts
+        unmatched_vol = sent_vol_from_to - sent_vol_to_from
+        # only use buffers, if both tokens are in the allowlist and buffer is sufficient to cover the trade
+        if not f in buffer_allow_listed_tokens or not t in buffer_allow_listed_tokens or buffers[t] < unmatched_vol:
+            sum_matched_vol += 2*min(sent_vol_from_to, sent_vol_to_from)
+            sum_rebalance_vol += unmatched_vol
+            nr_of_external_trades += 1
+        else:
+            buffers[f] += unmatched_vol
+            buffers[t] -= unmatched_vol
+            sum_matched_vol += unmatched_vol + 2 * \
+                min(sent_vol_from_to, sent_vol_to_from)
```

Comment:
`min(sent_vol_from_to, sent_vol_to_from)` is the matched_vol, right? Maybe define it as a variable above so it is clear.

Reply:
Ah, that is bad naming. `unmatched_vol` is the unmatched volume from COW and     `sum_matched_vol` is the matched volume from buffers. Let me fix it.